### DISCUSSION TITLE
USERDEFINED and ObjectType

### DISCUSSION THREAD

**Discussion by BIMster008** [buildingSMART/IDS#367] [Q&A] [answered]
Hi,
here goes my first post here.
I am trying to define an IDS for one project and to actually learn it more in detail. One aspect, that I still do not comprehend, is the ObjectType definition. I have read the documentation and seen the examples, but it is not clear to me whether the ObjectType is defined in the applicability or as a requirement? I could not find more accurate explanations, or at least that I could understand :)

I have tried to place the ObjectType in the applicability, but Bonsai is giving me an error "unexpected child". I have also tried it as a requirement in the property, but the result was faulty (no elements checked).
In my project, we have some USERDEFINED ObjectTypes, that I would like to include in the IDS.

I would appreciate it if someone could explain this method to me and maybe give an example.
Highly appreciated.

**Comment by Moult** [comment]
If you share some code that might help explain exactly what you're doing.

You just need to use the `predefinedType` filter of the entity facet. Anything you use that that isn't part of the standard enumeration will auto check USERDEFINED values for you.

**Comment by BIMster008** [comment]
Hi Dion,
on the example that I am working, I am talking about IfcWall.USERDEFINED:WALLSYSTEM
here are the codes that I have tried

Trying as part of an entity
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ids:ids xmlns:ids="http://standards.buildingsmart.org/IDS" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://standards.buildingsmart.org/IDS http://standards.buildingsmart.org/IDS/1.0/ids.xsd">
    <ids:info>
        <ids:title>IDS Check</ids:title>
        <ids:copyright>Name</ids:copyright>
		<ids:description>Title</ids:description>
    </ids:info>
    <ids:specifications>
            <ids:specification ifcVersion="IFC4" name="WALLS" description="Anforderungen für Wand">
            <ids:applicability minOccurs="0" maxOccurs="unbounded">
                <ids:entity>
                    <ids:name>
                        <ids:simpleValue>IFCWALL</ids:simpleValue>
                    </ids:name>
                    <ids:predefinedType>
                        <ids:simpleValue>USERDEFINED</ids:simpleValue>
                    </ids:predefinedType>
                    <ids:ObjectType>
                        <xs:enumeration value="WALLSYSTEM"/>
                    </ids:ObjectType>
                </ids:entity>
            </ids:applicability>
            <ids:requirements>
                <ids:property dataType="IFCLABEL">
                    <ids:propertySet>
                        <ids:simpleValue>TVG_Spezifisch</ids:simpleValue>
                    </ids:propertySet>
                    <ids:baseName>
                        <ids:simpleValue>Bauabschnitt</ids:simpleValue>
                    </ids:baseName>
                </ids:property>
              </ids:requirements>
        </ids:specification>
    </ids:specifications>
</ids:ids>
```



Trying is a property:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ids:ids xmlns:ids="http://standards.buildingsmart.org/IDS" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://standards.buildingsmart.org/IDS http://standards.buildingsmart.org/IDS/1.0/ids.xsd">
    <ids:info>
        <ids:title>IDS Check</ids:title>
        <ids:copyright>Name</ids:copyright>
		<ids:description>Title</ids:description>
    </ids:info>
    <ids:specifications>
            <ids:specification ifcVersion="IFC4" name="WALLS" description="Anforderungen für Wand">
            <ids:applicability minOccurs="0" maxOccurs="unbounded">
                <ids:entity>
                    <ids:name>
                        <ids:simpleValue>IFCWALL</ids:simpleValue>
                    </ids:name>
                    <ids:predefinedType>
                        <ids:simpleValue>USERDEFINED</ids:simpleValue>
                    </ids:predefinedType>
                </ids:entity>
            </ids:applicability>
            <ids:requirements>
                <ids:property dataType="IFCLABEL">
                    <ids:propertySet>
                        <ids:simpleValue>TVG_Spezifisch</ids:simpleValue>
                    </ids:propertySet>
                    <ids:baseName>
                        <ids:simpleValue>Bauabschnitt</ids:simpleValue>
                    </ids:baseName>
                    **<ids:value>
                        <xs:restriction base="xs:string">
                            <xs:enumeration value="WALLSYSTEM"/>
                        </xs:restriction>
                    </ids:value**>
                </ids:property>
              </ids:requirements>
        </ids:specification>
    </ids:specifications>
</ids:ids>
```

I am thankful for your help

**Comment by atomczak** [comment]
Hi @BIMster008, the ObjectType is an attribute, not a property. That's a very IFC-ish thing which has to be defined differently in the IDS. Here is the corrected applicability part:

```
  <ids:applicability minOccurs="0" maxOccurs="unbounded">
      <ids:entity>
          <ids:name>
              <ids:simpleValue>IFCWALL</ids:simpleValue>
          </ids:name>
          <ids:predefinedType>
              <ids:simpleValue>USERDEFINED</ids:simpleValue>
          </ids:predefinedType>
      </ids:entity>
      <ids:attribute>
          <ids:name>
              <ids:simpleValue>ObjectType</ids:simpleValue>
          </ids:name>
          <ids:value>
              <ids:simpleValue>WALLSYSTEM</ids:simpleValue>
          </ids:value>
      </ids:attribute>
  </ids:applicability>
```

**Reply by BIMster008** [reply]
@atomczak 
in the first moment, I thought it would be working. While testing it in Bonsai, I am getting a positive report, but no Elements have been checked (HTML result).
I also did a Test with a door with a userdefined status.

Here is the code, after I copied your suggestion:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ids:ids xmlns:ids="http://standards.buildingsmart.org/IDS" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://standards.buildingsmart.org/IDS http://standards.buildingsmart.org/IDS/1.0/ids.xsd">
    <ids:info>
        <ids:title>IDS Check</ids:title>
        <ids:copyright>Name</ids:copyright>
		<ids:description>Title</ids:description>
    </ids:info>
    <ids:specifications>
            <ids:specification ifcVersion="IFC4" name="WALLS" description="Anforderungen für abg. Decken">
            <ids:applicability minOccurs="0" maxOccurs="unbounded">
                <ids:entity>
                    <ids:name>
                        <ids:simpleValue>IFCDOOR</ids:simpleValue>
                    </ids:name>
                    <ids:predefinedType>
                        <ids:simpleValue>USERDEFINED</ids:simpleValue>
                    </ids:predefinedType>
                </ids:entity>
                <ids:attribute>
                    <ids:name>
                         <ids:simpleValue>ObjectType</ids:simpleValue>
                    </ids:name>
                    <ids:value>
                        <ids:simpleValue>INSPECTIONACCESS</ids:simpleValue>
                    </ids:value>
                </ids:attribute>
            </ids:applicability>
            <ids:requirements>
                <ids:property dataType="IFCLABEL">
                    <ids:propertySet>
                        <ids:simpleValue>TVG_Spezifisch</ids:simpleValue>
                    </ids:propertySet>
                    <ids:baseName>
                        <ids:simpleValue>Bauabschnitt</ids:simpleValue>
                    </ids:baseName>
                </ids:property>
                <ids:property dataType="IFCLABEL">
                    <ids:propertySet>
                        <ids:simpleValue>TVG_Spezifisch</ids:simpleValue>
                    </ids:propertySet>
                    <ids:baseName>
                        <ids:simpleValue>Bauteilname</ids:simpleValue>
                    </ids:baseName>
                </ids:property>             
            </ids:requirements>
        </ids:specification>
    </ids:specifications>
</ids:ids>

```

**Reply by Moult** [reply, marked as answer]
You don't need a separate attribute check. IDS works the same way as IFC, the ObjectType (or ElementType) is interpreted as the value of the predefined type.

```
  <ids:applicability minOccurs="0" maxOccurs="unbounded">
      <ids:entity>
          <ids:name>
              <ids:simpleValue>IFCDOOR</ids:simpleValue>
          </ids:name>
          <ids:predefinedType>
              <ids:simpleValue>INSPECTIONACCESS</ids:simpleValue>
          </ids:predefinedType>
      </ids:entity>
  </ids:applicability>
```

This is documented here: https://github.com/buildingSMART/IDS/blob/development/Documentation/ImplementersDocumentation/TestCases/entity/pass-a_predefined_type_may_specify_a_user_defined_element_type.ids and https://github.com/buildingSMART/IDS/blob/development/Documentation/ImplementersDocumentation/TestCases/entity/pass-a_predefined_type_may_specify_a_user_defined_element_type.ifc

**Reply by atomczak** [reply]
Cool! I didn't know that 😊 However, shouldn't both solutions work and find the same elements?

**Reply by BIMster008** [reply]
hi @Moult 
thank you for the explanation. To my surprise, this worked perfectly on Bonsai.
To be honest, I was not expecting this solution. I have read other sources, and also I have tried generating an IDS file from BIMcollab Zoom. BIMcollab also gave the same code as @atomczak recommended.
I apologize, but I feel a bit confused of what the actual solution here the correct one is?! The positive side is, that I can implement the code now for my project.

**Reply by Moult** [reply]
This solution is how IFC works. IFC's EXPRESS schema is defined such that the PredefinedType can only hold an enumeration, but not arbitrary text. So the user defined value has to be stored in another attribute (e.g. ObjectType). The fact that the user defined attribute is separate is just a detail, but not the design intent.

IFC also has rules around PredefinedType inheritance (which many vendors get wrong). For example, if an IfcDoorType is an INSPECTIONACCESS, then the IfcDoor automatically inherits that and is _also_ an INSPECTIONACCESS and each IfcDoor must _not_ have an ObjectType (i.e. it must be null).

For this reason, the more verbose solution by @atomczak is not completely correct, because it does not consider inheritance of the PredefinedType, nor does it consider that element types (if you happen to select them) have the attribute name of ElementType, and process types use the attribute ProcessType.

**Reply by BIMster008** [reply]
it's an honor to learn something new and happy that the system is working
Thanx @Moult 
Thanx @atomczak 

cheers,
agron
